Add `refresh` method to trigger a lookup call.

ithinkicancode opened the pull request:
First attempt to resolve #42.

ithinkicancode commented on zio-cache/shared/src/main/scala/zio/cache/Cache.scala:
```diff
@@ -188,19 +190,96 @@ object Cache {
+                        MapValue.Pending(key, promise),
+                        completedResult
+                      )
+                      map.put(k, refreshInProgress)
```
We might want to compare `value` with what `map.put(k, refreshInProgress)` returns. If another thread might have updated the map with `MapValue.Complete` already so here we might trigger another re-fetch/computation. In terms of correctness, I think it's still correct but ideally I would like to eliminate the possibility of duplicating the work.

adamgfraser replied:
Yes we need to use `replace` here to put the new value in the map only if the new value equals the old value.

ithinkicancode replied:
Fixed. Thanks Adam.

ithinkicancode commented on project/plugins.sbt:
```diff
@@ -9,10 +9,10 @@ addSbtPlugin("com.typesafe"                      % "sbt-mima-plugin"
 addSbtPlugin("org.portable-scala"                % "sbt-scala-native-crossproject" % "1.0.0")
 addSbtPlugin("org.portable-scala"                % "sbt-scalajs-crossproject"      % "1.0.0")
-addSbtPlugin("org.scala-js"                      % "sbt-scalajs"                   % "1.6.0")
+addSbtPlugin("org.scala-js"                      % "sbt-scalajs"                   % "1.7.0")
```
Had to update this otherwise JS tests wouldn't run.

ithinkicancode commented on build.sbt:
```diff
@@ -36,13 +36,12 @@ addCommandAlias(
   .in(file("."))
   .settings(
-    publish / skip := true,
-    unusedCompileDependenciesFilter -= moduleFilter("org.scala-js", "scalajs-library")
```
Had to remove this after upgrading `sbt-scalajs` to 1.7.0.

ithinkicancode commented:
A couple of to-do's:

- [ ] Update docs (new `refresh` method and the CacheStats)
- [ ] Add more tests

ithinkicancode commented:
Hi @adamgfraser, beyond the main difference between `get` and `refresh` methods, I realized that, when it comes to failure caching, the current implementation of `refresh` doesn't take previous failure into account - that is, when user calls `refresh`, it will do it. I can see both pros and cons. For now, at least to me, `refresh` is meant for "internal use", e.g. not exposed as an API as `get` would be. It's up to the developer's discretion how often and under what circumstance that `refresh` should be called. However, users can do whatever they want :) Besides that, it'd be a gotcha (or even a potential performance issue) if users are not aware of this difference when they do care of failure caching.

I can refactor to make sure we honor failure caching if you think that's the right way to go. Alternately, I can add a note to the documentation to highlight this difference. Let me know your thoughts on this. Thanks.

ithinkicancode commented:
Also, if you have any reservation to the current implementation, feel free to voice your concerns :) I'm happy to discuss and make the necessary changes.

adamgfraser commented on zio-cache/shared/src/main/scala/zio/cache/Cache.scala:
```diff
@@ -186,59 +174,149 @@ object Cache {
+          override def refresh(k: Key): IO[Error, Unit] =
+            ZIO.effectSuspendTotal {
+              for {
+                ref    <- Ref.make(map.get(k))
```
There is no need for a `Ref` here since everything we are doing is on a single thread.

ithinkicancode replied:
@adamgfraser Is it because the entire block of `ZIO.effectSuspendTotal` runs on a single thread? I looked at its impl and java docs don't mention that.

adamgfraser replied:
There isn't anything about `effectSuspendTotal` that does that you're just not doing anything concurrent here.

ithinkicancode replied:
Uh, I ran into some issues during the initial implementation. That was why I create a new promise (the question you asked below: https://github.com/zio/zio-cache/pull/52#discussion_r719054611), as opposed to the `var promise`... I'll revert to `var`'s (just like the `get` method) and make sure things are fine.

adamgfraser replied:
I think the problem there was we had the race condition because you were forking a fiber to do the computation of the new value versus just computing it synchronously.

ithinkicancode replied:
Yeah possibly. I'll spend some time today to make sure things will be kosher. Thanks.

adamgfraser commented on zio-cache/shared/src/main/scala/zio/cache/Cache.scala:
```diff
@@ -186,59 +174,149 @@ object Cache {
+                    map.remove(key, currentValue)
+                    get(key)
+                  } else {
+                    val refreshPromise    = newPromise()
```
Why are we creating a new promise here?

ithinkicancode replied:
Fixed

adamgfraser reviewed (changes requested):
Looks good. Left some comments below.

ithinkicancode commented on zio-cache/shared/src/main/scala/zio/cache/Cache.scala:
```diff
@@ -1,29 +1,24 @@
- * the previous result of the lookup function if it is available or else
- * compute a new result with the lookup function, put it in the cache, and
- * return it.
+ * A `Cache` is defined in terms of a lookup function that, given a key of type `Key`, can either fail with an error of
```
Sure thing. I upgraded scalafmt but the latest version does some funky stuff. I actually had to downgrade it back to 3.0.3 (instead of the latest 3.0.5).... Anyhow, I will figure out the format config for comments and bring the old style back :)

ithinkicancode commented on zio-cache/shared/src/test/scala/zio/cache/CacheSpec.scala:
```diff
@@ -68,17 +68,45 @@ object CacheSpec extends DefaultRunnableSpec {
+    ),
+    suite("`refresh` method")(
+      testM("should update the cache with a new value") {
+        var modifier = 1
```
Yup, will be part of my next commit.

ithinkicancode commented:
> Looks good. Left some comments below.

Hey @adamgfraser, I've addressed all your comments. If you get a chance, please have another review. Thanks.
